support/cas-server-support-thymeleaf-core/src/main/java/org/apereo/cas/services/web/CasThymeleafTemplatesDirector.java

```diff
@@ -33,6 +33,20 @@ public String getUrlExternalForm(final URL url) {
         return url.toExternalForm();
     }
 
+    /**
+     * Gets the URL external form supplemented by a question mark or an ampersand.
+     *
+     * @param url the URL
+     * @return the supplemented URL external form
+     */
+    public String getSupplementedUrlExternalForm(final URL url) {
+        val externalForm = url.toExternalForm();
+        if (externalForm.contains("?")) {
+            return externalForm + "&";
+        }
+        return externalForm + "?";
+    }
```

Comment:
I would suggest that you modify the original function to handle this scenario, instead of a new one.

Also, this check `if (externalForm.contains("?")) {` is slightly strange. Would it make sense to see if the URL just ends with a `?` ?

Reply:
In fact, the original function is already used in the `casPropagateLogoutView.html` to send the logout as an AJAX call:

```javascript
    $.ajax({
        url: [[${@casThymeleafTemplatesDirector.getUrlExternalForm(entry.key.logoutUrl)}]],
        dataType: 'jsonp',
        async: true,
        contentType: [[${entry.value.contentType}]]
        , data: [[${entry.value.message}]]
```

In that case, the additionnal `?` or `&` is not necessary as the message is not sent in the request URL.

Though, I can supplement the original function and not create a new one.

About the test, it really is a `contains` and not only a `endsWith` as the logout URL can be for example: `http://localhost:8080/cas/login?client_name=XXX`.

Reply:
Thanks for the notes. "Fixing" the original function sounds better to me. It seems less code to manage and test and regardless of where it's used `getUrlExternalForm` should just behave more correctly.

Reply:
OK. I changed it